services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/ZetaGatewayClient.java

```diff
@@ -0,0 +1,229 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.futurewei.alcor.common.db.CacheException;
+import com.futurewei.alcor.common.entity.ResponseId;
+import com.futurewei.alcor.common.enumClass.StatusEnum;
+import com.futurewei.alcor.common.executor.AsyncExecutor;
+import com.futurewei.alcor.dataplane.cache.VpcGatewayInfoCache;
+import com.futurewei.alcor.dataplane.client.grpc.DataPlaneClientImpl;
+import com.futurewei.alcor.dataplane.entity.MulticastGoalState;
+import com.futurewei.alcor.dataplane.entity.UnicastGoalState;
+import com.futurewei.alcor.dataplane.entity.ZetaPortGoalState;
+import com.futurewei.alcor.schema.Common;
+import com.futurewei.alcor.web.entity.dataplane.InternalPortEntity;
+import com.futurewei.alcor.web.entity.gateway.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+
+@Service("zetaGatewayClient")
+public class ZetaGatewayClient {
+    private static final Logger LOG = LoggerFactory.getLogger(DataPlaneClientImpl.class);
+
+    @Autowired
+    private VpcGatewayInfoCache gatewayInfoCache;
+
+    @Value("${zetaGateway.check.timeout}")
+    private String zetaGatewayCheckTimeout;
+
+    @Value("${zetaGateway.check.interval}")
+    private String zetaGatewayCheckInterval;
+
+    @Value("${microservices.zeta.service.url}")
+    private String zetaGatewayUrl;
+
+    @Value("${microservices.gateway.service.url}")
+    private String gatewayUrl;
+
+    private RestTemplate restTemplate = new RestTemplate();
+
+    public ZetaGatewayClient() {
+
+    }
+
+    public void updateVPCZetaGateway(GatewayInfo gatewayInfo) throws Exception {
+        String url = gatewayUrl + "project/" + "/gatewayinfo/" + gatewayInfo.getResourceId();
+        HttpEntity<GatewayInfoJson> gatewayHttpEntity = new HttpEntity<>(new GatewayInfoJson(gatewayInfo));
+        restTemplate.put(url, gatewayHttpEntity, ResponseId.class);
+    }
+
+    public void enableZetaGatewayForPort(InternalPortEntity portEntity) throws Exception {
+        GatewayInfo gatewayInfo = gatewayInfoCache.findItem(portEntity.getVpcId());
+        if (gatewayInfo == null) {
+            List<GatewayEntity> newGatewayEntities = new ArrayList<>();
+            newGatewayEntities.add(new GatewayEntity(null, null, null, null, GatewayType.ZETA,
+                    StatusEnum.NOTAVAILABLE.getStatus(), null, null, null, null, null, null));
+            GatewayInfo newGatewayInfo = new GatewayInfo(portEntity.getVpcId(), newGatewayEntities, null, "available");
+            gatewayInfoCache.addItem(newGatewayInfo);
```

Comment:
It seems that an empty Zeta GW is inserted to the cache. Do we really need this?

Reply:
If the code reach to this point, that's mean zetagateway is enabled. If there is no gatewayinfo in the dpm's cache for the VPC, that's mean VPC registration in the GM is not done yet or something wrong in the GM. We should create an empty zeta gateway for this vpc and indicate it's zeta gateway status = failed. Otherwise, the following port creation for this VPC has no clue to use zeta or ovs path.

Reply:
I see. Consider to log error or warning in this case.

Reply:
log.error added